lib/logstash/inputs/gelf.rb

```diff
@@ -68,36 +76,135 @@ def initialize(params)
   public
   def register
     require 'gelfd'
+    @tcp = nil
+    @udp = nil
   end # def register
 
   public
   def run(output_queue)
     begin
-      # udp server
-      udp_listener(output_queue)
+      if @use_tcp
+        if @port_tcp == 0
+          @port_tcp = @port
+        end
+        tcp_thr = Thread.new(output_queue) do |output_queue|
+          tcp_listener(output_queue)
+        end
+      end
+      if @use_udp
+        if @port_udp == 0
+          @port_udp = @port
+        end
+        udp_thr = Thread.new(output_queue) do |output_queue|
+          udp_listener(output_queue)
+        end
+      end
     rescue => e
       unless stop?
         @logger.warn("gelf listener died", :exception => e, :backtrace => e.backtrace)
         Stud.stoppable_sleep(RECONNECT_BACKOFF_SLEEP) { stop? }
         retry unless stop?
       end
     end # begin
+    if @use_tcp
+      tcp_thr.join
+    end
+    if @use_udp
+      udp_thr.join
+    end
   end # def run
 
   public
   def stop
-    @udp.close
+    if @use_tcp && @use_udp
+      return @udp.close && tcp.close
```

Comment:
We can replace this with:

```ruby
def close
  @udp.close if @use_udp
  @tcp.close if @use_tcp
rescue IOError
end
```

Reply:
Won't this affect the return value of the method and therefore change the behaviour of:
`       unless stop?
         @logger.warn("gelf listener died", :exception => e, :backtrace => e.backtrace)
         Stud.stoppable_sleep(RECONNECT_BACKOFF_SLEEP) { stop? }
         retry unless stop?`

Reply:
That is a fair question, however the method `stop?` [just checks if the `do_stop` method was called](https://github.com/elastic/logstash/blob/master/logstash-core/lib/logstash/inputs/base.rb#L94), so the return here of `stop` isn't relevant.

Reply:
I replaced the code as you suggested but you also renamed the method to close.
Where is this method ever used? all calls to stop always have a ? at the end and so they don't refer to this stop method right?

Reply:
Uh...I meant stop not close..
Apologies, this is what happens when doing or reviews on mobile,

Reply:
Ok. The way that I understand it now: stop and stop? are two different methods.

Inside the plugin stop? is called to find out if execution was aborted.
From outside the plugin stop is called to stop execution.

As a non ruby programmer stop and is_stopped would have been 1000x more clear to me :) I was wondering how this code even works

Reply:
Done